### PR TITLE
Addition of Goal and DisplayName config features, plus option ObfuscateProgressiveAchievements

### PR DESCRIPTION
**Pull request by alchemistmatt**
We are again looking at adopting this plugin for use on AddstarMC.  I have some improvements that I plan to add, but I see that you're actively working on this too.  I haven't got to my improvements yet, but I did see instances of code that could be cleaned up for readability.  The two commits in this pull request are those improvements.  I figured I should get these changes sent to you for approval before we start changing overlapping code.   None of the changes in these commits affects functionality; just code readability.  I have compiled and tested these changes.

### CODE REVIEW

**Comment by alchemistmatt** [comment]
There are now several more commits adding new functionality.  The achievements in the config file now support `Goal` and `DisplayName` entries, which define the text shown to the user in the GUI and when they obtain an achievement.  `DisplayName` allows the `Name` to define the official achievement name in the database, while `DisplayName` is a user-friendly name, and could be changed in the future without affecting any completed achievements.  `Goal` is the text shown for achievements that have not yet been obtained.

Example:
```yaml
  smooth_brick:
    '500':
      Goal: 'Place 500 stone bricks.'
      Message: '&4You have placed 500 stone bricks!'
      Name: place_500_smooth_brick
      DisplayName: 'Stone Brick Layer'
    '2500':
      Goal: 'Place 2500 stone bricks.'
      Message: '&4You have placed 2500 stone bricks!'
      Name: place_2500_smooth_brick
      DisplayName: 'Stone Brick Architect'
    '10000':
      Goal: 'Place 10000 stone bricks.'
      Message: '&4You have placed 10000 stone bricks!'
      Name: place_10000_smooth_brick
      DisplayName: 'Castle Builder'
````
The new config file option `ObfuscateProgressiveAchievements` applies at times when we have a series of related achievements (like the `smooth_brick` example above) where the only thing changing is the number of times the event has occurred.  With `ObfuscateProgressiveAchievements` set to true, if the player has placed 600 stone bricks, the first two achievements will be visible but the achievement with a threshold of 10000 bricks will be obfuscated in the GUI.
This change also applies to achievements that do not have subcategories, for example:
```yaml
 Fish:
   '10':
     Goal: 'Catch 10 fish.'
     Message: '&4You have caught 10 fish!'
     Name: fish_count_10
     DisplayName: 'Just getting started'
   '50':
     Goal: 'Catch 50 fish.'
     Message: '&4You have caught 50 fish!'
     Name: fish_count_50
     DisplayName: 'Perfect Casting'
   '100':
     Goal: 'Catch 100 fish.'
     Message: '&4You have caught 100 fish!'
     Name: fish_count_100
     DisplayName: 'Master Angler'
   '500':
     Goal: 'Catch 500 fish.'
     Message: '&4You have caught 500 fish!'
     Name: fish_count_500
     DisplayName: 'Trawler'
```
This allows us to have some fun achievement names, yet the player won't see the name until they are working toward that achievement.  So, if the player has caught 15 fish, `Just getting started` and `Perfect Casting` will be visible in the GUI, but `Master Angler` and `Trawler` will be obfuscated

**Comment by PyvesB** [comment]
Hello,

Thank you for your interest in the project. These seem like very nice additions, I like the ideas of having "progressive" obfuscation and separating Goal and DisplayName parameters. As far as the cleanup is concerned, it is welcome as well; the commits have been quite chaotic lately, with me wanting to complete the work for the long delayed next major update. I'm planning on reinforcing the comments and readability of the code in the upcoming days, so don't struggle too much with trying to improve my messy code parts for now. ;-)

I do have some concerns for the Name parameter though; you mentioned that it would be a database name; so if I have understood correctly, this would be some sort of unique achievement ID in the database, linking to the config. Nevertheless, the current format of the achievements table is different, and changing it would impact users if there is not some sort of database upgrade procedure to deal with it. 

The design of the table might seem a little clumsy, but it is done like this for historical reasons; the old "Achievement" plugin I based this project on was designed that way, and in the early days of the plugin, there were no such things as Commands achievements, and some users wanted to add some achievements to the config just on Christmas day for instance. As the list command did not exist at the time, the books could be simply generated solely with the database, regardless of whether the user made changes to the config or not. But the list command has since then broken this "independence" between the database and the config file.

Some users also use this big self contained table to display information on their website in a very simple manner. Having to link their website with the config file would be a pain. Sadly, I have no idea how many users use such a website interface.

To sum up, the current design of the table is not great, but things must be done with caution and by minimising the impact to current users as much as possible. I'm not merging right now to enable some extra discussion on this topic, although all the rest seems great to me!

Thanks again for your contributions!

Cheers,

Pyves

**Comment by alchemistmatt** [comment]
Thanks for getting back to me, and glad to see you welcome the changes.  You misunderstand my comments on `Name` and `DisplayName`.  The proposed changes do not alter what is stored in the database (or read from the database).  The achievement name stored in the database is the same `Name` as it has always been.  What's new is `DisplayName`.  If a `DisplayName` is defined, that text is shown to the user when they use `/aach list` and `/aach book` and when they receive an achievement: the large Title that is shown for a few seconds while the fireworks go off is `DisplayName`, but `Name` is stored in the achievements table in the database.

**Comment by PyvesB** [comment]
Indeed, I had misunderstood, it now makes sense when having a closer look into the code. All good! :-)

**Comment by PyvesB** [comment]
Hello again,

I noticed that we did have an issue when it comes to /aach book. As this command directly queries the database to get the information about the received achievements in a chronological order, we would  end up with the "Name" parameters inside it. Generating a book with the official database names such as "fish_count_10" inside wouldn't be great. How do you think we could simply address this concern?

**Comment by alchemistmatt** [comment]
I thought that I had already implemented a fix for that; perhaps I missed a spot.  See
 https://github.com/PyvesB/AdvancedAchievements/commit/a2e79a960ed8f932c6c026fc159bcc056226bf59#diff-ad2fe3b4e41a1bc00540dd61caee4fbbR1234 and 
https://github.com/PyvesB/AdvancedAchievements/commit/a2e79a960ed8f932c6c026fc159bcc056226bf59#diff-28785b2a3f271aa71ebabfa9e6c0a275R635

**Comment by PyvesB** [comment]
You're absolutely right, I had checked in BookCommand.java, but I missed that out. Perfect! ;-)
